File: src/main/java/com/medicalsystem/model/value/SelectFieldValue.java
package com.medicalsystem.model.value;

import com.medicalsystem.model.field.SelectField;

import javax.persistence.*;
import java.util.Map;

@Entity
@Table(name = "SELECT_FIELD_VALUES")
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
public class SelectFieldValue extends FieldValue<SelectField, String> {

    @Access(AccessType.PROPERTY)
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn
    @Override
    public SelectField getField() {
        return super.getField();
    }

    @Access(AccessType.PROPERTY)
    @Override
    public String getValue() {
        Map<String, String> possibleValues = getField().getPossibleValues();
        String selectedKey = super.getValue();
        return possibleValues.get(selectedKey);
    }

}

File: src/main/kotlin/com/medicalsystem/model/value/SelectFieldValue.kt
package com.medicalsystem.model.value

import com.medicalsystem.model.report.ReportField
import org.apache.poi.ss.usermodel.Cell
import javax.persistence.*

@Entity
@Table(name = "SELECT_FIELD_VALUES")
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
class SelectFieldValue : FieldValue<String?>() {

    @Column(name = "VALUE")
    override var value: String? = ""

    override fun setValueFromString(s: String) {
        this.value =
                if (field?.possibleValues!!.contains(s)) {
                    s
                } else {
                    null
                }
    }

    fun setValueByMapValue(mapValue: String) {
        field?.let {
            it.possibleValues.forEach { key, value ->
                if (value == mapValue)
                    this.value = key
            }
        }
    }

    override fun createCellValue(cell: Cell) {
        cell.setCellValue(value)
    }

    override fun fullfills(reportField: ReportField): Boolean {
        return false
    }
}